File: backend/src/models/post.ts
/**
 * Model purpose
 *
 * This model represents articles that user can write. userId is the author's id
 * who will create that post. The content should be in markdown format. This model
 * can be used to write about recipes as the content field be used for markdown
 * text
 */

import { Document, model, Schema } from "mongoose";
import { TagDocument } from "./tag";
import { UserDocument } from "./user";
import MongoPaging from "mongo-cursor-pagination";

export type PostDocument = Document & {
  title: string;
  description: string;
  content: string; // markdown
  readTime: number;
  wordCount: number;
  published: boolean;
  coverImgURL: string;
  userId: UserDocument;
  tags: TagDocument[];
};

const PostSchema = new Schema<PostDocument>(
  {
    title: {
      type: String,
      required: true,
      trim: true,
      maxlength: 2048,
      index: true,
    },
    description: {
      type: String,
      required: true,
      trim: true,
      maxlength: 4096,
      index: true,
    },
    content: { type: String, required: true, trim: true, index: true },
    readTime: { type: Number, required: true },
    wordCount: { type: Number, required: true },
    published: { type: Boolean, required: true, default: false },
    coverImgURL: {
      type: String,
      required: true,
      default: process.env.DEFAULT_USER_PROFILE_PIC_URL,
    },
    userId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    tags: {
      type: [{ type: Schema.Types.ObjectId, ref: "Tag" }],
      required: true,
    },
  },
  { timestamps: true }
);

// Pagination
PostSchema.plugin(MongoPaging.mongoosePlugin, { name: "paginatePost" });

PostSchema.index({ title: "text", description: "text", content: "text" });
const Post = model<PostDocument>("Post", PostSchema);
Post.createIndexes();

export default Post;

File: flutter-app/lib/models/post/post.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:salt/models/tag/tag.dart';
import 'package:salt/models/user/user.dart';

part 'post.g.dart';

@JsonSerializable()
class Post {
  final String id;
  final String title;
  final String description;
  final String content;
  final double readTime;
  final int wordCount;
  final bool published;
  final String coverImgURL;
  final User user;
  final List<Tag> tags;
  final DateTime updatedAt;
  final DateTime createdAt;

  Post({
    required this.id,
    required this.title,
    required this.description,
    required this.content,
    required this.readTime,
    required this.wordCount,
    required this.published,
    required this.coverImgURL,
    required this.user,
    required this.tags,
    required this.updatedAt,
    required this.createdAt,
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    return _$PostFromJson(json);
  }

  Map<String, dynamic> toJson() => _$PostToJson(this);
}
